### PR TITLE
Return ipfs origins for document.location.origin

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/13303

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)
- [ ] Requested a security/privacy review as needed

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by spylogsster** [comment]
Public gateway:
image

Local node:
image

File: chromium_src/third_party/blink/renderer/platform/weborigin/security_origin.cc [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+#include "third_party/blink/renderer/platform/wtf/text/string_builder.h"
+#include "third_party/blink/renderer/platform/wtf/text/wtf_string.h"
+
+namespace {
+
+const char* kIpfs[] = {"ipfs", ".ipfs.localhost"};
+const char* kIpns[] = {"ipns", ".ipns.localhost"};
+
+bool BuildRawIPFSIfApplicable(const String& host,
```
Comment (iefremov) [inline chromium_src/third_party/blink/renderer/platform/weborigin/security_origin.cc:14]: The new string representation appears to be insecure (as opposed to `http://hash.ipfs.localhost`), so callers of `ToString()` could get confused. From quick searching there are `ServiceWorkerContainer::registerServiceWorker`, `IsInCrossOriginIframe` in password autofill agent or `ResourceRequestHead::SetHTTPOrigin`. 

I'd prefer to have one of the following 
 * add a comment explainig why the change is safe for us
 * do less invasive change, e.g. modify only `document.location`, not `SecurityOrigin`
 * investigate more detailed approach, e.g. adding `ipfs` to secure schemes. Also note that `.ipfs.localhost` plays well with eTLD+1 checking, while `ipfs://hash` does not

Comment (spylogsster) [reply chromium_src/third_party/blink/renderer/platform/weborigin/security_origin.cc]: `do less invasive change, e.g. modify only document.location, not SecurityOrigin`
done

File: patches/third_party-blink-renderer-core-frame-location.cc.patch [hunk 2]
```diff
@@ -0,0 +1,12 @@
+diff --git a/third_party/blink/renderer/core/frame/location.cc b/third_party/blink/renderer/core/frame/location.cc
+index 4b344295426052d07fe5a66dde3b0b50f5f6d9c7..ec319be39eb9f41399cfc3fb6656c74b3dc622e9 100644
+--- a/third_party/blink/renderer/core/frame/location.cc
++++ b/third_party/blink/renderer/core/frame/location.cc
+@@ -95,6 +95,7 @@ String Location::search() const {
+ }
+ 
+ String Location::origin() const {
```
Comment (iefremov) [inline patches/third_party-blink-renderer-core-frame-location.cc.patch:8]: I gather we'd better patch `Location::Url()` instead? To change all the properties. 
also would be gread to add a test querying all of them from js

**Comment by diracdeltas** [comment]
in the past, we've had security issues when using URL parsers that did not fully agree on what the origin of a URL is. so i have two concerns:

1. internally the security origin is `http://$CID.ipfs.localhost:$PORT` so we would have to assume that there's no internal code which uses document.origin instead. (which could be true, but i haven't checked.) for instance, if some component injected scripts based on document.origin matching `http://...`, then this component wouldn't work correctly on IPFS pages.
2. external developers could create security bugs for themselves if the location.origin isn't consistent with `location.href` or other APIs. for example if a website does the following:


```
if (URL.parse(document.location.href).origin.endsWith('.localhost')) {
  user.isBanned = true
}

if (!document.location.origin.endsWIth('.localhost)) {
  user.isBanned = false
}
```

developers would expect `isBanned` to be true on `http://something.ipfs.localhost`, but it would be false with this change.

as mentioned in https://github.com/brave/brave-browser/issues/13303#issuecomment-781640166, after talking to lidel i'm not sure this change is worth doing, at least not without completely switching the internal origin to `ipfs://...` so everything is consistent.

**Comment by bbondy** [comment]
@iefremov can you advise on how we could go about actually switching things so that the network stack actually works with ipfs:// and it proxies things to the local gateway?

**Comment by bbondy** [comment]
I think we should probably abandon this approach

**Comment by iefremov** [comment]
@bbondy Is this more about making `ipfs` a first-class citizen inside renderer than network stack? 

We've discussed it a little bit with @spylogsster in DM: good starting point would be auditing `blink::SecurityOrigin` and teaching it to properly recognize `ipfs://`. The other two things to research are `blink::SchemeRegistry` and `registry_controlled_domain.h`. As Sergey noted, we already properly compare ipfs urls with `SameDomainOrHost`, but there can be other peculiarities in other functions.

Maybe there is a chance for upstreaming this work, since there is a lot happening around IPFS in upstream, e.g. https://blog.ipfs.io/2021-01-15-ipfs-and-igalia-collaborate-on-dweb-in-browsers/ (though I'm pretty sure you are well aware about it)

**Pull request closed without merging**